full-node/sov-stf-runner/src/prover_service/parallel/prover.rs

```diff
@@ -0,0 +1,178 @@
+use std::ops::Deref;
+use std::sync::{Arc, Mutex};
+
+use serde::de::DeserializeOwned;
+use serde::Serialize;
+use sov_rollup_interface::da::{BlockHeaderTrait, DaSpec};
+use sov_rollup_interface::services::da::DaService;
+use sov_rollup_interface::stf::StateTransitionFunction;
+use sov_rollup_interface::zk::{Proof, ZkvmHost};
+
+use super::{Hash, ProverServiceError};
+use crate::{ProofGenConfig, ProofSubmissionStatus, StateTransitionData};
+
+enum ProverStatus<StateRoot, Witness, Da: DaSpec> {
+    WitnessSubmitted(StateTransitionData<StateRoot, Witness, Da>),
+    Proving,
+    Proved(Proof),
+    Err(anyhow::Error),
+}
+
+#[derive(Default)]
+struct ProverState<StateRoot, Witness, Da: DaSpec> {
+    prover_status: HashMap<Hash, ProverStatus<StateRoot, Witness, Da>>,
+}
+
+impl<StateRoot, Witness, Da: DaSpec> ProverState<StateRoot, Witness, Da> {
+    fn remove(&mut self, hash: &Hash) -> Option<ProverStatus<StateRoot, Witness, Da>> {
+        self.prover_status.remove(hash)
+    }
+
+    fn set_to_proving(&mut self, hash: Hash) -> Option<ProverStatus<StateRoot, Witness, Da>> {
+        self.prover_status.insert(hash, ProverStatus::Proving)
+    }
+
+    fn set_to_proved(
+        &mut self,
+        hash: Hash,
+        proof: Result<Proof, anyhow::Error>,
+    ) -> Option<ProverStatus<StateRoot, Witness, Da>> {
+        match proof {
+            Ok(p) => self.prover_status.insert(hash, ProverStatus::Proved(p)),
+            Err(e) => self.prover_status.insert(hash, ProverStatus::Err(e)),
+        }
+    }
+
+    fn get_prover_status(&self, hash: Hash) -> Option<&ProverStatus<StateRoot, Witness, Da>> {
+        self.prover_status.get(&hash)
+    }
+}
+
+pub(crate) struct Prover<StateRoot, Witness, Da: DaService> {
+    prover_state: Arc<Mutex<ProverState<StateRoot, Witness, Da::Spec>>>,
+}
+
+impl<StateRoot, Witness, Da> Prover<StateRoot, Witness, Da>
+where
+    Da: DaService,
+    StateRoot: Serialize + DeserializeOwned + Clone + AsRef<[u8]> + Send + Sync + 'static,
+    Witness: Serialize + DeserializeOwned + Send + Sync + 'static,
+{
+    pub(crate) fn new() -> Self {
+        Self {
+            prover_state: Arc::new(Mutex::new(ProverState {
+                prover_status: Default::default(),
+            })),
+        }
+    }
+
+    pub(crate) fn submit_witness(
+        &self,
+        state_transition_data: StateTransitionData<StateRoot, Witness, Da::Spec>,
+    ) {
+        let header_hash = state_transition_data.da_block_header.hash().into();
+        let data = ProverStatus::WitnessSubmitted(state_transition_data);
+
+        self.prover_state
+            .lock()
+            .expect("Lock was poisoned")
+            .prover_status
+            .insert(header_hash, data);
+    }
+
+    pub(crate) fn start_proving<Vm, V>(
+        &self,
+        block_header_hash: Hash,
+        config: Arc<ProofGenConfig<V, Da, Vm>>,
+        mut vm: Vm,
+        zk_storage: V::PreState,
+    ) -> Result<(), ProverServiceError>
```

Comment:
Do you mean that the caller should spawn a separate thread or wait for the prove method to complete?

Reply:
No, I think we should indeed delegate the thread management to the prover service (as we are doing now).

However, spawning one thread per user request is analogous to having the user spawning his own threads. We can name this as `lazy` approach. The main problem of this approach is that the number of concurrent threads will be defined by the number of simultaneous requests, and this might deplete the resources of a server in case it faces a spike of requests.

What I'm suggesting is the future possibility of implementation of a `eager` load approach. Consider the following sketch:

![image](https://github.com/Sovereign-Labs/sovereign-sdk/assets/8730839/5469ecc9-6389-4086-aca1-8a41c0e432b0)

Here the user submits a STF to a prover service, that will generate him a oneshot channel. Such prover will then forward the work to be done (i.e. prove the STF) to a "work management unit" (in other words, a load balancer), that will pick an available worker thread and submit the `(stf, oneshot)` pair.

Provided the user can configure a "number of maximum workers", we end with a couple of advantages:

- The resource consumption of the architecture is predictable and is just the required resources for a single STF function, multiplied by the number of workers
- We don't build a queue of proved STFs waiting to be fetched. If the user never fetches such proofs, they will just stay in memory, growing indefinitely. Instead, if the user never requests for the proof, he will have at some point to drop the ownership of his oneshot channel - and that would free the resources allocated for the unused proof
- We preload the worker threads, removing the overhead of initiating a new thread per request
- We will end up having lower latency times as no CPU cycle is wasted on waiting for the user to request his proof. They are always computed when submitted, as the `eager` approach

We might not need to make this entire change now, but we might profit from some infrastructure unrelated to the STF itself such as a configuration argument with the maximum number of threads (to avoid denial of service). As next step, we can introduce this `eager` work management

Reply:
discussed offline

Reply:
As discuss, we might embed the parallelism arguments `ProofGenConfig` for now, and maybe change that once the fork management implementation is completed